### PR TITLE
Add new content for Apps Drivers

### PR DESCRIPTION
**Pull request by neflyte**
#### Summary
Create new content for the Apps Drivers topic; Golang, specifically.
Content will be updated in KV store, OAuth2 store, App webhooks, and subscriptions topics.
The App testing topic will be removed.
The Apps in the Marketplace topic will be redirected to the plugin marketplace doc

Subtopics include:

- [x] Installation
- [x] Manifests
- [x] Bindings and forms
- [x] Call handlers
- [x] Static assets
- [x] Full code example
- [x] Helper + utility functions

Functionality docs enhancements:
- [x] K/V store
- [x] OAuth2 store
- [x] Subscriptions
- [x] App webhooks

### CODE REVIEW

File: site/content/integrate/apps/drivers/golang.md
Comment (cwarnermm) [inline site/content/integrate/apps/drivers/golang.md:10]:
```suggestion
Install the driver using `go get`. The version number `v1.1.0` in the command below  can be updated to reference newer releases.
```

**Review by cwarnermm — APPROVED** [review]
One small piece of feedback. Approving for expediency. Great job! This is very easy to read and follow.

**Comment by asatkinson** [comment]
Are the equivalent rest api versions of OAuth2 calls going to be in this PR, or is there a separate PR for those?

**Comment by asatkinson** [comment]
KV-Store functionality needs to talk about how the bot token is used and how each user (including the bot user) has its own namespace.  So for example, you can have one key "name" and each user will have a separate version of it in the KV-Store including the bot-user

File: site/content/integrate/apps/functionality/oauth2.md
Comment (mickmister) [inline site/content/integrate/apps/functionality/oauth2.md:47]:
```suggestion
The `/oauth2/connect` and `/oauth2/complete` calls can be customized using the `get_oauth2_connect_url` and `on_oauth2_complete` properties of the App's manifest, respectively.
```

File: site/content/integrate/apps/functionality/oauth2.md
Comment (mickmister) [inline site/content/integrate/apps/functionality/oauth2.md:152]: This below syntax will safely handle both cases of these cases:

- `callRequest.Values["client_secret"]` is `nil`
- `callRequest.Values["client_secret"]` is not a `string`

```suggestion
	clientSecret, ok := callRequest.Values["client_secret"].(string)
	if !ok {
		// handle the error
	}
```

Comment (mickmister) [reply site/content/integrate/apps/functionality/oauth2.md:152]: There's also a `callRequest.GetValue("client_secret", "")` method that does this type assertion, though the dev still needs to handle the empty string case.

File: site/content/integrate/apps/functionality/oauth2.md
Comment (mickmister) [inline site/content/integrate/apps/functionality/oauth2.md:131]: Do the docs ever define/reference the form being used for this submission?

File: site/content/integrate/apps/functionality/oauth2.md
Comment (mickmister) [inline site/content/integrate/apps/functionality/oauth2.md:268]: https://github.com/mattermost/mattermost-plugin-apps/pull/397

```suggestion
The user's access token can also be obtained from the call request context `oauth2_user` field. The call must be configured to expand the `oauth2_user` context field with expand level `all`.
```

File: site/content/integrate/apps/functionality/kv-store.md
Comment (mickmister) [inline site/content/integrate/apps/functionality/kv-store.md:10]: I don't understand what this means. The App will have access to the user or bot token when it wants to access this configuration right? For HTTP-based apps, they can store that information elsewhere if they want to. In the case where something occurs outside of the ecosystem of Mattermost (such as webhook call, or some other non-Mattermost event/trigger), the App framework's webhook endpoint should be used to "poke" the App with the bot token.

I suppose my question is, what is this "App configuration" being mentioned here, that would in some scenario make since to _not_ store in the KV store?

File: site/content/integrate/apps/functionality/kv-store.md
Comment (mickmister) [inline site/content/integrate/apps/functionality/kv-store.md:22]: The driver is technically using HTTP endpoints, so this is a little confusing. It seems like we're saying it's two completely separate concepts.

```suggestion
- Using HTTP REST endpoints directly
- Using a [driver]({{<ref "/integrate/apps/drivers">}})
```

File: site/content/integrate/apps/functionality/kv-store.md
Comment (mickmister) [inline site/content/integrate/apps/functionality/kv-store.md:94]: It should be noted that `myData` can be any type that can be unmarshaled from JSON. The `KVGet` method automatically does the unmarshaling for you.

File: site/content/integrate/apps/functionality/kv-store.md
Comment (mickmister) [inline site/content/integrate/apps/functionality/kv-store.md:154]: This example shows a more complex data type, so maybe this already handles the explanation of my previous comment

**Review by mickmister — COMMENTED** [review]
Awesome work @neflyte! There's a lot of great content here.

I have a few suggestions and some questions on the changes

File: site/content/integrate/apps/authentication/_index.md
Comment (neflyte) [reply site/content/integrate/apps/authentication/_index.md:9]: No. This section will be entirely rewritten.

File: site/content/integrate/apps/drivers/_index.md
Comment (neflyte) [reply site/content/integrate/apps/drivers/_index.md:15]: @asatkinson Please confirm the above. :)

File: site/content/integrate/apps/drivers/_index.md
Comment (neflyte) [reply site/content/integrate/apps/drivers/_index.md:7]: You don't need to implement the client with a driver. If you don't have a driver, you have to "implement" the App REST endpoints and the MM REST endpoints.

Is there another way we could describe drivers such that they're attractive to implementers?

File: site/content/integrate/apps/drivers/golang.md
Comment (neflyte) [reply site/content/integrate/apps/drivers/golang.md:206]: The first code example on the page shows the import path for the `httputils` package.

File: site/content/integrate/apps/functionality/oauth2.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/oauth2.md:268]: @mickmister What does the `summary` level give? Should it not also give this information? Is there a `summary` level for this expand field at all?

File: site/content/integrate/apps/functionality/kv-store.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/kv-store.md:22]: They are to an implementor. One's easy, the other's low-level.
@asatkinson Please confirm if we should have a driver concept at all given the REST endpoints we've been documenting.

File: site/content/integrate/apps/functionality/kv-store.md
Comment (mickmister) [reply site/content/integrate/apps/functionality/kv-store.md:10]: Stating the fact that an appropriate access token is needed seems to be good, though I think this is mentioned elsewhere when we talk about the user-based scoping in the store. Yeah I think we can remove this note

File: site/content/integrate/apps/functionality/oauth2.md
Comment (mickmister) [reply site/content/integrate/apps/functionality/oauth2.md:268]: Currently there is nothing provided for `summary`. From https://github.com/mattermost/mattermost-plugin-apps/pull/397#issuecomment-1323555279:

> Coming back to this change: 2/5 that I like the current behaviour as it's'more strict and allows us to add a meaningful use of `summary` later.

File: site/content/integrate/apps/drivers/golang.md
Comment (azigler) [reply site/content/integrate/apps/drivers/golang.md:6]: Yes, I agree with @mickmister, I think it's important to tell them that step. If this were a Node project, we'd probably also manually tell them to `npm init` because that signifies to the reader what we're about to do. I think we should do the same here.

**Comment by asatkinson** [comment]
@mickmister, do you have any further comments for @neflyte or should he be good to cleanup your existing comments and ship this, or do you have more files to go through?

@neflyte there are a lot of good details in @mickmister's comments around exact parameters and small example code changes.  If these don't work a user will have difficulty, so please do follow his lead.

In general though this is enormously better than what we had before.  Great work!

File: site/content/integrate/apps/functionality/oauth2.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/oauth2.md:268]: Why does this code include the `summary` level in its check, then?
https://github.com/mattermost/mattermost-plugin-apps/blob/0856d8f8eac70040d6715f7c3733a3d3ca288ddc/server/proxy/expand.go#L138

File: site/content/integrate/apps/drivers/golang.md
Comment (neflyte) [reply site/content/integrate/apps/drivers/golang.md:228]: Does the "Static Assets" page not cover that point, or should we also include it here?

File: site/content/integrate/apps/functionality/oauth2.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/oauth2.md:131]: No. The example golang code that I took it from does, however. I can link that code here to give further context.

I'd like to avoid including forms in this page since it's primarily a reference on how to use the OAuth2 endpoints/functionality. Forms are described on a different page. That doesn't mean, however, that we can't include a reminder or note about the form needing a particular field or whatnot.

**Review by levb — APPROVED** [review]

File: site/content/integrate/apps/functionality/oauth2.md
Comment (mickmister) [reply site/content/integrate/apps/functionality/oauth2.md:268]: Please see the code on the master branch https://github.com/mattermost/mattermost-plugin-apps/blob/2e3c62c9a83a2a5c899576bd5db78922164acd14/server/proxy/expand.go#L141

Comment (mickmister) [reply site/content/integrate/apps/functionality/oauth2.md:268]: Any discussion about this particular issue can be discussed here where there is already discussion captured https://github.com/mattermost/mattermost-plugin-apps/pull/397

File: site/content/integrate/apps/functionality/oauth2.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/oauth2.md:30]: Expand level `summary` will also work.

File: site/content/integrate/apps/functionality/oauth2.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/oauth2.md:268]: We can only document released functionality in this PR.

https://github.com/mattermost/mattermost-plugin-apps/pull/397 has the label "Docs/Not Needed". Should it have the "Docs/Needed" label instead?

Comment (mickmister) [reply site/content/integrate/apps/functionality/oauth2.md:268]: The reason it has "Docs/Not Needed" is because the PR was closed. The "Docs/Needed" label is on the PR that introduced the changes however https://github.com/mattermost/mattermost-plugin-apps/pull/336

Comment (mickmister) [reply site/content/integrate/apps/functionality/oauth2.md:268]: > We can only document released functionality in this PR.

I don't want to have outdated information that doesn't work though. We should assume 1.2 is released since we are releasing it very soon to resolve issues like this

Comment (mickmister) [reply site/content/integrate/apps/functionality/oauth2.md:268]: @levb @hanzei Thoughts on including documentation that relates to v1.2 of the framework in this PR?

File: site/content/integrate/apps/drivers/_index.md
Comment (asatkinson) [reply site/content/integrate/apps/drivers/_index.md:7]: Two options:
(1) Something like this

_Mattermost provides API libraries, called **drivers**, that wrap the Mattermost REST APIs to make it easier to for a developer to make calls to the Mattermost Server._

(2) Eliminate this drivers section and talk about it in context of the "Mattermost API" section like this PR I whipped up
https://github.com/mattermost/mattermost-developer-documentation/pull/1185

File: site/content/integrate/apps/functionality/oauth2.md
Comment (cwarnermm) [reply site/content/integrate/apps/functionality/oauth2.md:256]: Our developer documentation should detail what's released and available publicly and should not include future functionality not yet released. Doing so can damage trust we're building up among our developer and integration communities. 

Please hold off on releasing 1.2 documentation until 1.2 has been released. CC @asatkinson

File: site/content/integrate/apps/drivers/_index.md
Comment (neflyte) [reply site/content/integrate/apps/drivers/_index.md:7]: I like a bit of both. :smile:  I've updated the Drivers page with the above statement, and also updated the "Mattermost API" page with the suggested content in #1185.

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md]: The `summary` level also gives the secret key.

File: site/content/integrate/apps/functionality/oauth2.md
Comment (mickmister) [reply site/content/integrate/apps/functionality/oauth2.md]: @cwarnermm In this case, there is a breaking change to something not fully fleshed out previously. This specific issue has significantly affected partnering orgs creating Apps and our internally developed Apps because of this detail. It's simply a "bug fix" I'm trying to put into the documentation here.

I understand the point of making documentation consistent with what's available, but 5/5 it's a moot point in this exact situation. I proposed that this breaking change may constitute a major version of the App framework, but the team decided not to so this, since the feature had not been flshed out yet https://github.com/mattermost/mattermost-plugin-apps/pull/397#issuecomment-1323555279. Please see the discussion on that PR for more details

File: site/content/integrate/apps/functionality/kv-store.md
Comment (azigler) [reply site/content/integrate/apps/functionality/kv-store.md:18]: I agree with these points!

**Review by mickmister — CHANGES_REQUESTED** [review]
Blocking on team discussion of these threads:

- https://github.com/mattermost/mattermost-developer-documentation/pull/1172#discussion_r1042601781
- https://github.com/mattermost/mattermost-developer-documentation/pull/1172#discussion_r1042486837
- https://github.com/mattermost/mattermost-developer-documentation/pull/1172#discussion_r1042487657

**Review by azigler — COMMENTED** [review]
Took a spin through this documentation and it makes sense to me. I think some of the documentation struggles with explaining REST vs using a driver (and how the latter is really just an abstracted, easy-to-use library for the former). The Golang driver page looks like it's trying to be exhaustive, but I noticed it's missing the calls for KV.

I think the Golang driver page would be easier to understand if it were divided into the same subsections as are under Functionality in the side menu, because then it would be a direct comparison and we could just give them driver examples corresponding to each REST-driven Functionality piece.

I learned a lot, thanks!

File: site/content/integrate/apps/functionality/kv-store.md
Comment (mickmister) [reply site/content/integrate/apps/functionality/kv-store.md:24]: Yeah that seems fine

**Comment by neflyte** [comment]
Per @mickmister, I will change the docs such that the following are true:
- In order to receive the webhook secret in App, you need to use `all`. For every other field of App, you can use `summary`.
- In order to receive any field of `OAuth2User` or `OAuth2App` you need to use `all`.

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (mickmister) [inline site/content/integrate/apps/functionality/external-webhooks/_index.md:42]: The framework automatically authenticates the webhook requests, so Apps should not be storing the secret for this purpose. That's covered in this other PR as well though so we can reconcile there if you'd like https://github.com/mattermost/mattermost-developer-documentation/pull/1184

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (mickmister) [inline site/content/integrate/apps/functionality/external-webhooks/_index.md:78]: The indentation looks off here

File: site/content/integrate/apps/structure/call/_index.md
Comment (mickmister) [inline site/content/integrate/apps/structure/call/_index.md:24]: Is this bracket here on purpose?

**Review by asatkinson — APPROVED** [review]
I like this a lot, but would like you to get approval from Ben/Michael on the smaller technical details.

Michael is going to look into what is required for REST calls with Embedded bindings.  We need to change the Embedded Bindings page, but this can come in a separate PR.

File: site/content/integrate/apps/structure/call/_index.md
Comment (neflyte) [reply site/content/integrate/apps/structure/call/_index.md:24]: Nope. That's something my mouse likes to leave around. ;)

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md:78]: How so? Can you provide a screenshot?

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md:42]: I will remove this note.

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md:16]: The default value for webhook security is no security at all, unless you're suggesting the default value is actually `secret`. Please confirm.

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md:39]: So how does the webhook secret mechanism even work? I'm getting more and more confused every time I touch this subject.

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (mickmister) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md]: Do you think we should note that the framework automatically checks the secret on webhook requests though?

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (mickmister) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md:16]: The default value is `secret`. I'm thinking we should have the secret in URL here

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (levb) [inline site/content/integrate/apps/functionality/external-webhooks/_index.md:92]: s/and will not perform any authentication against the call request./and will pass all incoming webhooks to the app's webhook call handler, unauthenticated. This may have security and hosting cost implications. /

**Review by levb — CHANGES_REQUESTED** [review]

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md]: What about the Authentication section? Does that also stay? It'd be redundant.
What are the downsides of not including it in this specific example?

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md]: This was discussed on a call with @asatkinson @mickmister and @levb; appropriate modifications to the above will be made.

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md]: I cannot reproduce that. Could you include a screenshot from your display showing this misalignment?

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md]: I've noted that as part of App webhook authentication.

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (mickmister) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md]: On the snippet I'm commenting on here

image

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (mickmister) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md]: Authentication section can stay. I just think the webhook URL is incomplete without having the secret in there. It shows that it should be included in the URL in order to be a usable webhook URL

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md]: Here's what I see:
image

I think this could be a display issue in your browser. Is the alignment also off in the Cnanges tab of the PR if whitespace is ignored?

Comment (azigler) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md]: I see what @mickmister is seeing here.

Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md]: Silly me. I removed this block completely a few commits ago. :) Nothing to see here!

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (mickmister) [inline site/content/integrate/apps/functionality/external-webhooks/_index.md:184]: If we set this json block to be `js` instead of `json`, this comment won't be highlighted in red here. The syntax highlighting may be a bit different though. Maybe we can omit this comment since it's not integral to the example?

File: site/content/integrate/apps/structure/bindings/embedded/_index.md
Comment (mickmister) [inline site/content/integrate/apps/structure/bindings/embedded/_index.md:84]: I think the HTTP example above is enough, and this non-driver Golang example may be overkill

**Review by mickmister — APPROVED** [review]
LGTM :tada: Just a couple non-blocking suggestions. No need to request re-review if you include any of those changes :+1:

File: site/content/integrate/apps/functionality/external-webhooks/_index.md
Comment (neflyte) [reply site/content/integrate/apps/functionality/external-webhooks/_index.md:184]: I would like to keep this here. I've used it in many places to indicate that there are more fields the implementer can expect in the output.

File: site/content/integrate/apps/structure/bindings/embedded/_index.md
Comment (asatkinson) [reply site/content/integrate/apps/structure/bindings/embedded/_index.md:84]: I'd agree with @mickmister - we probably only need one http example and it should be the non-golang one.  A Golang developer would probably use the driver which is documented below.

There is a screenshot of the golang driver example.  Do we have an equivalent screenshot for the coffee roast example?

Comment (neflyte) [reply site/content/integrate/apps/structure/bindings/embedded/_index.md]: Yes. I've updated the screenshots and examples, and I have removed the Golang HTTP example.

Comment (asatkinson) [reply site/content/integrate/apps/structure/bindings/embedded/_index.md]: Looks good.

**Pull request merged**
